api/src/test/java/com/redhat/insights/ITReportControllerTest.java

```diff
@@ -0,0 +1,288 @@
+/* Copyright (C) Red Hat 2023 */
+package com.redhat.insights;
+
+import static org.awaitility.Awaitility.await;
+import static org.junit.jupiter.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mock;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.json.JsonMapper;
+import com.redhat.insights.config.InsightsConfiguration;
+import com.redhat.insights.doubles.*;
+import com.redhat.insights.jars.ClasspathJarInfoSubreport;
+import com.redhat.insights.jars.JarInfo;
+import com.redhat.insights.jars.JarInfoSubreport;
+import com.redhat.insights.logging.InsightsLogger;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.*;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import org.junit.jupiter.api.Test;
+
+public class ITReportControllerTest {
+  private static final InsightsLogger logger = new NoopInsightsLogger();
+
+  /** Test there are not reports sent, if opted out */
+  @Test
+  public void testOptedOut() throws InterruptedException {
+    StoringInsightsHttpClient httpClient = new StoringInsightsHttpClient();
+    InsightsReport report = mock(InsightsReport.class);
+    InsightsConfiguration config = MockInsightsConfiguration.ofOptedOut("test_app");
+
+    InsightsReportController instance =
+        InsightsReportController.of(logger, config, report, () -> httpClient);
+
+    // when trying to generate report, it should throw an exception, if opted out
+    assertThrows(InsightsException.class, instance::generate);
+
+    // give controller some time to send a report, if it wants
+    // but no reports should be sent
+    Thread.sleep(TimeUnit.SECONDS.toSeconds(2));
```

Comment:
We're using [Awaitility](http://www.awaitility.org/) in some other tests, it's always better to use it than a hard-coded `Thread.sleep`

Reply:
This is tricky. I changed Thread.sleep to Awaitility where it made sense. Thing is, in these test scenarios test want to give controller time to send a report, but no report should actually be send (there is something other blocking sending it).
Se basically here we want to wait and check that nothing happened. I don't see a way how awaility can really help in that.
I can change the code to something like `await().atLeast(2 seconds).checkSomething` but I don't really see a benefit in doing so.